src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -27,30 +30,37 @@ internal sealed class ConsulConfigurationProvider : ConfigurationProvider
 
             _consulConfigClient = consulConfigClient;
             _source = source;
+            _cancellationTokenSource = new CancellationTokenSource();
 
             if (source.ReloadOnChange)
             {
-                ChangeToken.OnChange(
-                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _source.CancellationToken),
+                _changeTokenRegistration = ChangeToken.OnChange(
+                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _cancellationTokenSource.Token),
                     async () =>
                     {
-                        await DoLoad(true).ConfigureAwait(false);
+                        await DoLoad(true, _cancellationTokenSource.Token).ConfigureAwait(false);
                         OnReload();
                     });
             }
         }
 
         public override void Load()
         {
-            DoLoad(false).GetAwaiter().GetResult();
+            DoLoad(false, _cancellationTokenSource.Token).GetAwaiter().GetResult();
         }
 
-        private async Task DoLoad(bool reloading)
+        public void Dispose()
+        {
+            _cancellationTokenSource?.Cancel();
```

Comment:
I think the `_cancellationTokenSource` should also be disposed here.

Reply:
It should. But we should wait for all of the users of it first. The plan is to have a poll task instead of the
change token and wait for the task then dispose the token source.